RA-40-Implement-RecipesList-Caching

DenisGubernatorov opened the pull request:
1. Реализация работы со списком рецептов из БД. 
2. Логику оставил как и для категорий. Если в БД есть данные, то берет оттуда. Если данных нет, то пытается грузить через API и, в случае успеха, записывает их в БД. Что-то в телеге скринкаст отослать не получилось. Но все работает штатно

[Запись экрана от 2025-03-17 09-30-29.webm](https://github.com/user-attachments/assets/0c9bffc3-76b1-4334-818d-2fdbc38f81e3)

3. И нюанс по списку рецептов. У нас в Recipe нет явного "указателя" в какую категорию он входит. Но, судя по приходящим id из api, за категорию отвечает разряд сотен. Поэтому получаю рецепты исходя из этой логике. Считаю что все рецепты вида 1xx относятс к категории с id 1. Для бургеров (id = 0) ,беру рецепты id  которых <100

4. Для упрощения, сохранение в БД полей ingredients и method сделал через сериализацию/десереализзацию, так как room не поддерживает сохранение не примитивов.  Хотя, по хорошему, для ингридиентов надо было бы вводить отдельную таблицу в БД

DenisGubernatorov committed:
RA-40-Implement-RecipesList-Caching

1) Реализация работы с БД для получения списка рецептов

DenisGubernatorov committed:
Merge branch 'master' into RA-40-Implement-RecipesList-Caching

danelyan reviewed (changes requested):
> 2. Логику оставил как и для категорий. Если в БД есть данные, то берет оттуда. Если данных нет, то пытается грузить через API и, в случае успеха, записывает их в БД. Что-то в телеге скринкаст отослать не получилось. Но все работает штатно

Попробуй почистить кеш, если база изначально пуста, то пишет ошибку.
Скринкаст можно не делать, мы перепроверили ))

> И нюанс по списку рецептов. У нас в Recipe нет явного "указателя" в какую категорию он входит. Но, судя по приходящим id из api, за категорию отвечает разряд сотен. Поэтому получаю рецепты исходя из этой логике. Считаю что все рецепты вида 1xx относятс к категории с id 1. Для бургеров (id = 0) ,беру рецепты id которых <100

Да, АПИ не всегда бывают совершенны )
Такой подход приемлем

> Для упрощения, сохранение в БД полей ingredients и method сделал через сериализацию/десереализзацию, так как room не поддерживает сохранение не примитивов. Хотя, по хорошему, для ингридиентов надо было бы вводить отдельную таблицу в БД

Да, отличное решение! Конечно, можно было бы сделать отдельную сущность в базе, но у нас нет отдельного экрана или сервиса для управления ингридиентами. Это был бы оверинжиниринг на текущем этапе

Давай поправим проблему пустой базы

DenisGubernatorov committed:
RA-40-Implement-RecipesList-Caching

1) Расширено логгирование при получении загрузки изображений, списков категорий и рецептов для проверки.
2) Поправлена ошибка, что если в списке избранного ничего нет, то показывало toast с ошибкой, а не сообщение о том, что список рецептов пуст.

DenisGubernatorov commented:
> Попробуй почистить кеш, если база изначально пуста, то пишет ошибку.
> Скринкаст можно не делать, мы перепроверили ))

Скринакаст я записывал к коменту https://github.com/DenisGubernatorov/RecipesApp/pull/35#discussion_r1997676883. Логика отрабатывает штатно. Сначала пытается загрузить данные из БД, а если их не удалось получить, то лезет в API и, при успешном получении данных, записывает их в БД.  На эмуляторе делал wipe data, а на телефоне удалял приложение полностью.
Нашел только ошибку,  что если список избранного пустой, то показывается не сообщение, что нет избранных рецептов,  а toast с ошибкой получения данных. Я максимально расширил логгирование при получении изображений, списков категорий и рецептов. Если ошибка связана с ними, а не со списком избранного, то просьба приложить логи. Отфильтровать по tag:RRE level:error tag:GlideException.

danelyan reviewed (approved):
Да, теперь отлично!

ps: Зайди еще в Инопланетный бургер. Тебе пригодится такой сниппет в следующем PR'е :) :
```kotlin
                try {
                    BigDecimal(quantityStr).multiply(BigDecimal(quantity))
                        .setScale(1, RoundingMode.HALF_UP)
                        .stripTrailingZeros()
                        .toPlainString()
                } catch (e: Exception) {
                    quantityStr
                }
```
